homeworks/week2/Minh-Le.ts

```diff
@@ -0,0 +1,32 @@
+//Repl url for testing: https://repl.it/join/svpniinp-minhd1
+
+class Main {
+  static main(args: string[]) {
+    const s: string[] = ["h", "e", "l", "l", "o", " ", "w", "o", "r", "l", "d"];
+    Main.removeDuplicateCharacters(s);
+    // s = ['h', 'e', 'l', 'o', ' ', 'w', 'r', 'd'];
+    console.log(s.join(""));
+    // Output: "helo wrd"
+  }
+
+  // Please comment on approaches you are taking, and their space and time complexity.
+  // If time is allowed, you can provide multiple versions of the method with trade-off in space and time complexity.
+  
+  // Time complexity of the method below: O(n), space complexity O(1) 
+  // since the size of map variable is upper-bounded by the number of possible characters 
+  static removeDuplicateCharacters(s: string[]): void {
+    const map = {};
+    for (let i = 0; i < s.length; i++) {
+      if (!map[s[i]]) {
+        map[s[i]] = i // store index of first-occurence letters into the map
+      }
+    }
+
+    // remove duplicated character by indexes from the end
+    for (let i = s.length - 1; i >= 0; i--) {
+      if (i !== map[s[i]]) { // check if the current index is the first-occurence index of the letter
+        s.splice(i, 1);
```

Comment:
This line removes a character from the input string, is it O(1)? If not, what is the time complexity of this algorithm?
You should add some tests also. It's always a plus when candidate does that. It also tells us that you have tested your code.

Reply:
the splice method in JavaScript has time complexity of O(1)

Reply:
It's O(n^2). The easiest way to estimate is to count the number of nested loops. Another way is to write some examples and walk through it.

In this specific exercise, the number of unique characters is bounded so you can argue that O(n^2) will come back down to O(n) when string length is greater than a certain threshold. However, if this is not a string but an array of integer, the time complexity will stay at O(n^2).

Reply:
https://stackoverflow.com/questions/22614237/javascript-runtime-complexity-of-array-functions/22615787